generators/javascript.js

```diff
@@ -16,305 +16,319 @@ import * as Variables from '../core/variables.js';
+  YIELD: 17,            // yield
+  COMMA: 18,            // ,
+  NONE: 99,             // (...)
+};
 
 /**
- * List of outer-inner pairings that do NOT require parentheses.
- * @type {!Array<!Array<number>>}
+ * JavaScript code generator class.
  */
-JavaScript.ORDER_OVERRIDES = [
-  // (foo()).bar -> foo().bar
-  // (foo())[0] -> foo()[0]
-  [JavaScript.ORDER_FUNCTION_CALL, JavaScript.ORDER_MEMBER],
-  // (foo())() -> foo()()
-  [JavaScript.ORDER_FUNCTION_CALL, JavaScript.ORDER_FUNCTION_CALL],
-  // (foo.bar).baz -> foo.bar.baz
-  // (foo.bar)[0] -> foo.bar[0]
-  // (foo[0]).bar -> foo[0].bar
-  // (foo[0])[1] -> foo[0][1]
-  [JavaScript.ORDER_MEMBER, JavaScript.ORDER_MEMBER],
-  // (foo.bar)() -> foo.bar()
-  // (foo[0])() -> foo[0]()
-  [JavaScript.ORDER_MEMBER, JavaScript.ORDER_FUNCTION_CALL],
+export class JavascriptGenerator extends CodeGenerator {
+  /**
+   * List of outer-inner pairings that do NOT require parentheses.
+   * @type {!Array<!Array<number>>}
+   */
+  ORDER_OVERRIDES = [
+    // (foo()).bar -> foo().bar
+    // (foo())[0] -> foo()[0]
+    [Order.FUNCTION_CALL, Order.MEMBER],
+    // (foo())() -> foo()()
+    [Order.FUNCTION_CALL, Order.FUNCTION_CALL],
+    // (foo.bar).baz -> foo.bar.baz
+    // (foo.bar)[0] -> foo.bar[0]
+    // (foo[0]).bar -> foo[0].bar
+    // (foo[0])[1] -> foo[0][1]
+    [Order.MEMBER, Order.MEMBER],
+    // (foo.bar)() -> foo.bar()
+    // (foo[0])() -> foo[0]()
+    [Order.MEMBER, Order.FUNCTION_CALL],
 
-  // !(!foo) -> !!foo
-  [JavaScript.ORDER_LOGICAL_NOT, JavaScript.ORDER_LOGICAL_NOT],
-  // a * (b * c) -> a * b * c
-  [JavaScript.ORDER_MULTIPLICATION, JavaScript.ORDER_MULTIPLICATION],
-  // a + (b + c) -> a + b + c
-  [JavaScript.ORDER_ADDITION, JavaScript.ORDER_ADDITION],
-  // a && (b && c) -> a && b && c
-  [JavaScript.ORDER_LOGICAL_AND, JavaScript.ORDER_LOGICAL_AND],
-  // a || (b || c) -> a || b || c
-  [JavaScript.ORDER_LOGICAL_OR, JavaScript.ORDER_LOGICAL_OR]
-];
+    // !(!foo) -> !!foo
+    [Order.LOGICAL_NOT, Order.LOGICAL_NOT],
+    // a * (b * c) -> a * b * c
+    [Order.MULTIPLICATION, Order.MULTIPLICATION],
+    // a + (b + c) -> a + b + c
+    [Order.ADDITION, Order.ADDITION],
+    // a && (b && c) -> a && b && c
+    [Order.LOGICAL_AND, Order.LOGICAL_AND],
+    // a || (b || c) -> a || b || c
+    [Order.LOGICAL_OR, Order.LOGICAL_OR]
+  ];
 
-/**
- * Whether the init method has been called.
- * @type {?boolean}
- */
-JavaScript.isInitialized = false;
+  constructor(name) {
+    super(name ?? 'JavaScript');
+    this.isInitialized = false;
 
-/**
- * Initialise the database of variable names.
- * @param {!Workspace} workspace Workspace to generate code from.
- */
-JavaScript.init = function(workspace) {
-  // Call Blockly.CodeGenerator's init.
-  Object.getPrototypeOf(this).init.call(this);
+    // Copy Order values onto instance for backwards compatibility
+    // while ensuring they are not part of the publically- -advertised
+    // API.
```

Comment:
Also if we do, can you add information to the PR description like you did for #7150 with before-and-after, and add the `deprecating change` label so it can be added to release notes?

Reply:
I think ideally we would, though there's no hurry.  I've added a TODO suggesting that when we do we replace them temporarily with a `set` accessor that calls `deprecate.warn`.  I considered adding such an accessor immediately but I don't think there's any urgency in getting developers to update—let's give them a chance to update their code _before_ we start spewing deprecation notices into the console.

Reply:
Agreed on not spamming them with deprecation notices immediately since there are going to be a LOT of them, but consider updating and tagging this PR anyway, because I think the release notes are going to be one of the main ways developers find out we want them to do this kind of update